In [31]:
import pandas as pd
import numpy as np

In [32]:
df = pd.read_csv('../data/synthetic_premium_bordereaux.csv')

In [33]:
df.shape

(6000, 69)

In [34]:
df.head(5)

,Total,BRANCH CODE,CLASS OF BUSINESS,PRODUCT CODE,PRODUCT DESC,TYPE OF COVER,USAGE TYPE,BUISSNESS DESC,POLICY FROM DATE,POLICY TO DATE,...,EFFECTIVE SI,RISK LOCATION,EQ ZONE,POLICY CREATOR,POLICY APPROVER,ENDORSEMENT DATE,ENTRY COMPLETE BY,ENTRY COMPLETE DT,CLAIM YTD PAID AMT,BUSINESS CLASS
0,105 - First Assurance-Kisumu,105,20,2010,Fire - Industrial,Comprehensive,Roads and Pipelines,New Business,26/01/2025,26/01/2026,...,10233715.49,Kisumu,Zone A,Robert J.,Jeffery W.,03/02/2025,Anthony G.,03/02/2025 00:00,0.0,Renewal Business
1,102 - First Assurance-Nakuru,102,60,1001,Terrorism and Political Violence (PVT),Comprehensive,MOTOR CHAUFFER DRIVEN,New Business,09/06/2025,09/06/2026,...,1043532.26,Nakuru,Zone A,Eric B.,Matthew D.,15/06/2025,Colin A.,15/06/2025 00:00,0.0,New Business
2,106 - First Assurance-Nyeri,106,20,2010,Fire - Industrial,Comprehensive,Others,New Business,25/01/2026,25/01/2027,...,13815706.69,Nyeri,Zone A,Christine H.,Barbara B.,02/02/2026,Brandon P.,02/02/2026 00:00,0.0,Renewal Business
3,106 - First Assurance-Nyeri,106,70,7003,ENG - Erection All Risk,Comprehensive,MOTOR CHAUFFER DRIVEN,Renewal Business,27/02/2025,27/02/2026,...,56150346.00,Nyeri,Zone C,Francisco K.,Elizabeth M.,06/03/2025,Frances J.,06/03/2025 00:00,0.0,Renewal Business
4,100 - First Assurance-Head Office,100,70,7003,ENG - Erection All Risk,Comprehensive,INSTITUTIONAL VEHICLES,New Business,18/02/2025,18/02/2026,...,5671503.44,Head Office,Zone A,Mark T.,Brian U.,21/02/2025,Linda M.,21/02/2025 00:00,0.0,Renewal Business


In [35]:
df.dtypes

Total                     str
BRANCH CODE             int64
CLASS OF BUSINESS       int64
PRODUCT CODE            int64
PRODUCT DESC              str
                       ...   
ENDORSEMENT DATE          str
ENTRY COMPLETE BY         str
ENTRY COMPLETE DT         str
CLAIM YTD PAID AMT    float64
BUSINESS CLASS            str
Length: 69, dtype: object

In [36]:
df.isnull().sum()

Total                 0
BRANCH CODE           0
CLASS OF BUSINESS     0
PRODUCT CODE          0
PRODUCT DESC          0
                     ..
ENDORSEMENT DATE      0
ENTRY COMPLETE BY     0
ENTRY COMPLETE DT     0
CLAIM YTD PAID AMT    0
BUSINESS CLASS        0
Length: 69, dtype: int64

In [37]:
df.dtypes

Total                     str
BRANCH CODE             int64
CLASS OF BUSINESS       int64
PRODUCT CODE            int64
PRODUCT DESC              str
                       ...   
ENDORSEMENT DATE          str
ENTRY COMPLETE BY         str
ENTRY COMPLETE DT         str
CLAIM YTD PAID AMT    float64
BUSINESS CLASS            str
Length: 69, dtype: object

# Data Cleaning

In [38]:
## Dropping irrelevant columns
df = df.drop(columns= ['UPR FM','UPR TO'])
df = df.drop(columns= ['ENDORSEMENT DATE'])
df =df.drop(columns=['ENTRY COMPLETE DT'])


In [39]:
import pandas as pd

# converting date columns to datetime
df ['POLICY FROM DATE'] = pd.to_datetime(df['POLICY FROM DATE'], format='%d/%m/%Y')
df['POLICY TO DATE'] = pd.to_datetime(df['POLICY TO DATE'], format='%d/%m/%Y')
df['APPROVAL DATE'] = pd.to_datetime(df['APPROVAL DATE'], format = '%d/%m/%Y')

In [40]:
df.head(5)

,Total,BRANCH CODE,CLASS OF BUSINESS,PRODUCT CODE,PRODUCT DESC,TYPE OF COVER,USAGE TYPE,BUISSNESS DESC,POLICY FROM DATE,POLICY TO DATE,...,NATURE OF BUSINESS,OCCUPATION,EFFECTIVE SI,RISK LOCATION,EQ ZONE,POLICY CREATOR,POLICY APPROVER,ENTRY COMPLETE BY,CLAIM YTD PAID AMT,BUSINESS CLASS
0,105 - First Assurance-Kisumu,105,20,2010,Fire - Industrial,Comprehensive,Roads and Pipelines,New Business,2025-01-26,2026-01-26,...,Motor Spares Shop,MARKETING,10233715.49,Kisumu,Zone A,Robert J.,Jeffery W.,Anthony G.,0.0,Renewal Business
1,102 - First Assurance-Nakuru,102,60,1001,Terrorism and Political Violence (PVT),Comprehensive,MOTOR CHAUFFER DRIVEN,New Business,2025-06-09,2026-06-09,...,Department Stores,Insurance Company,1043532.26,Nakuru,Zone A,Eric B.,Matthew D.,Colin A.,0.0,New Business
2,106 - First Assurance-Nyeri,106,20,2010,Fire - Industrial,Comprehensive,Others,New Business,2026-01-25,2027-01-25,...,"Hypermarket, Super markets, Shopping Malls, Co...",Not Working / Retired,13815706.69,Nyeri,Zone A,Christine H.,Barbara B.,Brandon P.,0.0,Renewal Business
3,106 - First Assurance-Nyeri,106,70,7003,ENG - Erection All Risk,Comprehensive,MOTOR CHAUFFER DRIVEN,Renewal Business,2025-02-27,2026-02-27,...,Department Stores,Businessman,56150346.00,Nyeri,Zone C,Francisco K.,Elizabeth M.,Frances J.,0.0,Renewal Business
4,100 - First Assurance-Head Office,100,70,7003,ENG - Erection All Risk,Comprehensive,INSTITUTIONAL VEHICLES,New Business,2025-02-18,2026-02-18,...,Chemical Plants (Except Petrochemical Industry),DIRECTORS,5671503.44,Head Office,Zone A,Mark T.,Brian U.,Linda M.,0.0,Renewal Business


In [41]:
df['APPROVAL DATE'].dtype
df['POLICY FROM DATE'].dtype
df['POLICY TO DATE'].dtype

dtype('<M8[us]')

In [42]:
df.columns.tolist()

['Total',
 'BRANCH CODE',
 'CLASS OF BUSINESS',
 'PRODUCT CODE',
 'PRODUCT DESC',
 'TYPE OF COVER',
 'USAGE TYPE',
 'BUISSNESS DESC',
 'POLICY FROM DATE',
 'POLICY TO DATE',
 'APPROVAL DATE',
 'DOC NO',
 'POLICY NO',
 'POLICY ENDORSEMENT NO',
 'OLD POLICY NO',
 'CUST CATEGORY CODE',
 'CUST CATEGORY DESC',
 'CUSTOMER CODE',
 'CUSTOMER NAME',
 'RISK NOTE NUMBER',
 'CUST PIN NO',
 'CUST ID NO',
 'INSURED CODE',
 'INSURED',
 'UW YEAR',
 'GROSS PREMIU',
 'BROKER COMM',
 'INW COMM',
 'RETN SI',
 'RETN',
 'CQS',
 'CQS COMM',
 'QUOTA SHARE SI',
 'QUOTA SHARE',
 'QS COMM',
 'SURPLUS 01 SI',
 'SURPLUS 01',
 'SURPLUS 01 COMM',
 'SURPLUS 02 SI',
 'SURPLUS 02',
 'SURPLUS 02 COMM',
 'FAC OUT',
 'FAC COMM',
 'SUM INSURED',
 'ACCOUNT TYPE',
 'COVER TYPE',
 'LOCATION',
 'BINDER CLASSIFICATION',
 'TTY APPR YN',
 'TTY PROCESS YN',
 'PROV FAC PERC',
 'FAC PLACED PERC',
 'UNPLACED FAC PERC',
 'FRONTING YN',
 'RISK CLASSIFICATION',
 'NATURE OF BUSINESS',
 'OCCUPATION',
 'EFFECTIVE SI',
 'RISK LOCATION',
 'E

In [43]:
columns_to_keep = ['BRANCH CODE', 'CLASS OF BUSINESS', 'PRODUCT DESC', 'TYPE OF COVER', 'USAGE TYPE', 'BUSINESS CLASS', 'POLICY FROM DATE', 'POLICY TO DATE', 'APPROVAL DATE', 'CUST CATEGORY DESC', 'UW YEAR', 'GROSS PREMIU', 'BROKER COMM', 'RETN', 'SUM INSURED', 'FAC OUT', 'RISK CLASSIFICATION', 'NATURE OF BUSINESS', 'CLAIM YTD PAID AMT']
df_clean=df[columns_to_keep]

In [44]:
df_clean.head(5)

,BRANCH CODE,CLASS OF BUSINESS,PRODUCT DESC,TYPE OF COVER,USAGE TYPE,BUSINESS CLASS,POLICY FROM DATE,POLICY TO DATE,APPROVAL DATE,CUST CATEGORY DESC,UW YEAR,GROSS PREMIU,BROKER COMM,RETN,SUM INSURED,FAC OUT,RISK CLASSIFICATION,NATURE OF BUSINESS,CLAIM YTD PAID AMT
0,105,20,Fire - Industrial,Comprehensive,Roads and Pipelines,Renewal Business,2025-01-26,2026-01-26,2025-02-03,Insurance Company,2025,281462.60,56292.52,237667.20,10233715.49,143827.56,Risk Category C -50%,Motor Spares Shop,0.0
1,102,60,Terrorism and Political Violence (PVT),Comprehensive,MOTOR CHAUFFER DRIVEN,New Business,2025-06-09,2026-06-09,2025-06-15,Tied Agents,2025,26489.79,2648.98,24065.88,1043532.26,0.00,Risk Category A -100%,Department Stores,0.0
2,106,20,Fire - Industrial,Comprehensive,Others,Renewal Business,2026-01-25,2027-01-25,2026-02-02,Insurance Company Brokers,2026,322426.93,32242.69,197700.08,13815706.69,0.00,Risk Category A -100%,"Hypermarket, Super markets, Shopping Malls, Co...",0.0
3,106,70,ENG - Erection All Risk,Comprehensive,MOTOR CHAUFFER DRIVEN,Renewal Business,2025-02-27,2026-02-27,2025-03-06,Direct/Assured,2025,984553.06,0.00,518455.51,56150346.00,0.00,Hazardous Risk - Ret 40%-20%,Department Stores,0.0
4,100,70,ENG - Erection All Risk,Comprehensive,INSTITUTIONAL VEHICLES,Renewal Business,2025-02-18,2026-02-18,2025-02-21,Agent,2025,202496.56,30374.48,153461.09,5671503.44,0.00,Prone to Major Losses - Ret 80%-60%,Chemical Plants (Except Petrochemical Industry),0.0


In [45]:
df_clean.columns.tolist()

['BRANCH CODE',
 'CLASS OF BUSINESS',
 'PRODUCT DESC',
 'TYPE OF COVER',
 'USAGE TYPE',
 'BUSINESS CLASS',
 'POLICY FROM DATE',
 'POLICY TO DATE',
 'APPROVAL DATE',
 'CUST CATEGORY DESC',
 'UW YEAR',
 'GROSS PREMIU',
 'BROKER COMM',
 'RETN',
 'SUM INSURED',
 'FAC OUT',
 'RISK CLASSIFICATION',
 'NATURE OF BUSINESS',
 'CLAIM YTD PAID AMT']

In [46]:
df_clean.shape
df_clean.dtypes

BRANCH CODE                     int64
CLASS OF BUSINESS               int64
PRODUCT DESC                      str
TYPE OF COVER                     str
USAGE TYPE                        str
BUSINESS CLASS                    str
POLICY FROM DATE       datetime64[us]
POLICY TO DATE         datetime64[us]
APPROVAL DATE          datetime64[us]
CUST CATEGORY DESC                str
UW YEAR                         int64
GROSS PREMIU                  float64
BROKER COMM                   float64
RETN                          float64
SUM INSURED                   float64
FAC OUT                       float64
RISK CLASSIFICATION               str
NATURE OF BUSINESS                str
CLAIM YTD PAID AMT            float64
dtype: object

In [47]:
df['FAC OUT'].dtype

dtype('float64')

In [48]:
df_clean['loss_ratio'] = df_clean['CLAIM YTD PAID AMT'] / df_clean['SUM INSURED']
df_clean.head(5)

,BRANCH CODE,CLASS OF BUSINESS,PRODUCT DESC,TYPE OF COVER,USAGE TYPE,BUSINESS CLASS,POLICY FROM DATE,POLICY TO DATE,APPROVAL DATE,CUST CATEGORY DESC,UW YEAR,GROSS PREMIU,BROKER COMM,RETN,SUM INSURED,FAC OUT,RISK CLASSIFICATION,NATURE OF BUSINESS,CLAIM YTD PAID AMT,loss_ratio
0,105,20,Fire - Industrial,Comprehensive,Roads and Pipelines,Renewal Business,2025-01-26,2026-01-26,2025-02-03,Insurance Company,2025,281462.60,56292.52,237667.20,10233715.49,143827.56,Risk Category C -50%,Motor Spares Shop,0.0,0.0
1,102,60,Terrorism and Political Violence (PVT),Comprehensive,MOTOR CHAUFFER DRIVEN,New Business,2025-06-09,2026-06-09,2025-06-15,Tied Agents,2025,26489.79,2648.98,24065.88,1043532.26,0.00,Risk Category A -100%,Department Stores,0.0,0.0
2,106,20,Fire - Industrial,Comprehensive,Others,Renewal Business,2026-01-25,2027-01-25,2026-02-02,Insurance Company Brokers,2026,322426.93,32242.69,197700.08,13815706.69,0.00,Risk Category A -100%,"Hypermarket, Super markets, Shopping Malls, Co...",0.0,0.0
3,106,70,ENG - Erection All Risk,Comprehensive,MOTOR CHAUFFER DRIVEN,Renewal Business,2025-02-27,2026-02-27,2025-03-06,Direct/Assured,2025,984553.06,0.00,518455.51,56150346.00,0.00,Hazardous Risk - Ret 40%-20%,Department Stores,0.0,0.0
4,100,70,ENG - Erection All Risk,Comprehensive,INSTITUTIONAL VEHICLES,Renewal Business,2025-02-18,2026-02-18,2025-02-21,Agent,2025,202496.56,30374.48,153461.09,5671503.44,0.00,Prone to Major Losses - Ret 80%-60%,Chemical Plants (Except Petrochemical Industry),0.0,0.0


In [49]:
df_clean['loss_ratio'].describe()

count    6000.000000
mean        0.008013
std         0.030026
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         0.415378
Name: loss_ratio, dtype: float64

In [50]:
df_clean['POLICY FROM DATE'] = pd.to_datetime(df_clean['POLICY FROM DATE'], format='%d/%m/%Y')
df_clean['POLICY YEAR']= df_clean['POLICY FROM DATE'].dt.year
df_clean['POLICY MONTH'] = df_clean['POLICY FROM DATE'].dt.month

In [51]:
df_clean[['POLICY FROM DATE', 'POLICY YEAR', 'POLICY MONTH']].head()

,POLICY FROM DATE,POLICY YEAR,POLICY MONTH
0,2025-01-26,2025,1
1,2025-06-09,2025,6
2,2026-01-25,2026,1
3,2025-02-27,2025,2
4,2025-02-18,2025,2


## 2. Exploratory Analysis

### 1. How many policies per branch
### 2. How many policies Per product line?
### 3. What does total gross premium look like by branch and by product?
### 4. What's the overall claims frequency (% of policies with a claim > 0) and average loss ratio, split by product line?
### 5. New vs renewal business — what's the mix, and does loss ratio differ between the two?
### 6. Any branch or product that stands out as unusually high loss ratio compared to the rest?






In [52]:
#1. How many policies per branch
df_clean.groupby('BRANCH CODE')['POLICY FROM DATE'].count()

BRANCH CODE
100    867
101    816
102    873
103    841
104    855
105    849
106    899
Name: POLICY FROM DATE, dtype: int64

In [53]:
#2 How many policies per product line
df_clean.groupby('PRODUCT DESC')['POLICY FROM DATE'].count()

PRODUCT DESC
Burglary                                  427
ENG - Erection All Risk                   381
ENG- Contractor All Risk                  402
ENG- Electronic Equipment                 402
Employers Liability                       411
Fire - Industrial                         414
Fire Domestic Package                     374
Marine Cargo                              400
Marine-Goods in Transit                   425
Marine-Open Cover                         404
Motor Commercial                          392
Motor Private                             396
Public Liability                          385
Terrorism and Political Violence (PVT)    396
WIBA Plus GPA                             391
Name: POLICY FROM DATE, dtype: int64

In [54]:
# 3. What does total gross premium look like by branch and by product?
df_clean.groupby(['BRANCH CODE', 'PRODUCT DESC'])['GROSS PREMIU'].sum()

BRANCH CODE  PRODUCT DESC                          
100          Burglary                                  2.123176e+07
             ENG - Erection All Risk                   1.452935e+07
             ENG- Contractor All Risk                  2.338682e+07
             ENG- Electronic Equipment                 1.670506e+07
             Employers Liability                       5.492093e+06
                                                           ...     
106          Motor Commercial                          5.616669e+06
             Motor Private                             4.346744e+06
             Public Liability                          5.689022e+06
             Terrorism and Political Violence (PVT)    3.193316e+06
             WIBA Plus GPA                             2.640796e+06
Name: GROSS PREMIU, Length: 105, dtype: float64

In [55]:
# 4 What's the overall claims frequency (% of policies with a claim > 0) and average loss ratio, split by product line?
# logic - first create  flag/helper column where 1 represents a policy with a claim while 0 represents a policy without a claim

df_clean['Has_Claim'] = df_clean['CLAIM YTD PAID AMT'] > 0

In [56]:
df_clean['Has_Claim'] = df_clean['Has_Claim'].astype(int)

In [57]:
df_clean['Has_Claim'].value_counts()

Has_Claim
0    5309
1     691
Name: count, dtype: int64

In [58]:
df_clean.groupby('PRODUCT DESC')[['Has_Claim', 'loss_ratio']].mean().sort_values('loss_ratio', ascending=False)

,Has_Claim,loss_ratio
PRODUCT DESC,,
ENG- Contractor All Risk,0.141791,0.010783
ENG - Erection All Risk,0.128609,0.009264
Fire Domestic Package,0.114973,0.008896
Marine-Open Cover,0.126238,0.008662
Burglary,0.121780,0.008542
ENG- Electronic Equipment,0.119403,0.008443
Motor Commercial,0.112245,0.008203
Public Liability,0.137662,0.008128
Fire - Industrial,0.111111,0.008007


### Engineering and industrial fire products show the highest average loss ratios, consistent with their exposure to large, infrequent losses, while motor and liability lines show more moderate but frequent claims" — a defensible, industry-plausible sentence without overclaiming precision from what's ultimately a synthetic sample.

In [59]:
# 5. New vs renewal business — what's the mix, and does loss ratio differ between the two?
df_clean.groupby('BUSINESS CLASS').agg(

    {
        'loss_ratio' : 'mean',
        'POLICY FROM DATE': 'count'

    }
)

,loss_ratio,POLICY FROM DATE
BUSINESS CLASS,,
New Business,0.008581,3051
Renewal Business,0.007426,2949


### From the exploratory analysis done specifically from the new vs renewal business done, it is worth noting that, the mix: 3051 New Business vs 2949 Renewal Business — roughly a 51%/49% split, so close to evenly balanced. Loss ratio comparison: New Business (0.0086) is only marginally higher than Renewal (0.0074), interesting observation that new business tends to run slightly higher loss ratios than renewals



## 3. Visualizations